src/GraphQL/Execution/SubscriptionExecutionStrategy.cs

```diff
@@ -0,0 +1,296 @@
+        /// Initializes a new instance with the specified execution strategy for child nodes.
+        /// </summary>
+        public SubscriptionExecutionStrategy(IExecutionStrategy baseExecutionStrategy)
+        {
+            _baseExecutionStrategy = baseExecutionStrategy ?? throw new ArgumentNullException(nameof(baseExecutionStrategy));
+        }
+
+        /// <summary>
+        /// Gets a static instance of <see cref="SubscriptionExecutionStrategy"/>.
+        /// </summary>
+        public static SubscriptionExecutionStrategy Instance { get; } = new();
+
+        /// <summary>
+        /// Executes a GraphQL subscription request and returns the result. The result consists
+        /// of one or more streams of GraphQL responses, returned within <see cref="SubscriptionExecutionResult.Streams"/>.
+        /// No serializable <see cref="ExecutionResult"/> is directly returned.
+        /// </summary>
+        public override async Task<ExecutionResult> ExecuteAsync(ExecutionContext context)
+        {
+            var rootType = GetOperationRootType(context);
+            var rootNode = BuildExecutionRootNode(context, rootType);
+
+            var streams = await ExecuteSubscriptionNodesAsync(context, rootNode.SubFields!).ConfigureAwait(false);
+
+            // if execution is successful, errors and extensions are not returned per the graphql-ws protocol
+            // if execution is unsuccessful, the DocumentExecuter will add context errors to the result
+
+            return new SubscriptionExecutionResult(context)
+            {
+                Executed = true,
+                Streams = streams,
+            };
+        }
+
+        private async Task<IDictionary<string, IObservable<ExecutionResult>>?> ExecuteSubscriptionNodesAsync(ExecutionContext context, ExecutionNode[] nodes)
+        {
+            var streams = new Dictionary<string, IObservable<ExecutionResult>>();
+
+            foreach (var node in nodes)
+            {
+                var stream = await ResolveEventStreamAsync(context, node).ConfigureAwait(false);
+                if (stream == null)
+                    return null;
+                streams[node.Name!] = stream;
+            }
+
+            return streams;
+        }
+
+        /// <summary>
+        /// Asynchronously returns a stream of <see cref="ExecutionResult"/> responses for the
+        /// specified <see cref="ExecutionNode"/>.
+        /// </summary>
+        protected virtual async Task<IObservable<ExecutionResult>?> ResolveEventStreamAsync(ExecutionContext context, ExecutionNode node)
+        {
+            context.CancellationToken.ThrowIfCancellationRequested();
+
+            var resolveContext = new ReadonlyResolveFieldContext(node, context);
+
+            IObservable<object?> subscription;
+
+            try
+            {
+                if (node.FieldDefinition?.Subscriber == null)
+                {
+                    // this is normally caught by schema validation
```

Comment:
Where? I don't see any `Subscriber` property usage from schema validation.

Reply:
Hmmm.  Perhaps I was thinking of your recent comment (while working on the AutoRegisteringObjectGraphType) that we should add validation in this respect.

Reply:
If you're planning to add such validation, I will leave the comment as-is.  I took a quick look at SchemaValidationVisitor, and I don't see a context, so I'm not exactly sure how you would add such validation.

Reply:
```suggestion
                    // todo: this should be caught by schema validation
```

Reply:
Here it is:
- #2994

Reply:
:) I was looking for it in "my issues", but I did not find it, obviously.